File: api/schemas/authenticate/authenticate.py
import graphene

from functions.input_validators import cleanse_input
from schemas.shared_structures import AuthResult
from schemas.authenticate.sign_in_user import sign_in_user
from scalars.email_address import EmailAddress


class Authenticate(graphene.Mutation):
    """
    This mutation allows users to give their credentials and retrieve a token
    that gives them access to restricted content.
    """

    # Define mutation arguments
    class Arguments:
        user_name = EmailAddress(
            description="User email that they signed up with.", required=True,
        )
        password = graphene.String(description="Users password", required=True,)

    # Define mutation fields
    auth_result = graphene.Field(
        lambda: AuthResult, description="User info who just signed in, and their JWT"
    )

    # Define mutation functionality
    @staticmethod
    def mutate(self, info, **kwargs):
        # Get arguments
        user_name = cleanse_input(kwargs.get("user_name"))
        password = cleanse_input(kwargs.get("password"))

        # Create user and JWT
        user_info = sign_in_user(user_name=user_name, password=password)

        # Return information to user
        return Authenticate(
            AuthResult(str(user_info.get("auth_token")), user_info.get("user"))
        )

File: api-js/src/mutations/user/authenticate.js
const { GraphQLNonNull, GraphQLString, GraphQLInt } = require('graphql')
const { mutationWithClientMutationId } = require('graphql-relay')
const { t } = require('@lingui/macro')
const { authResultType } = require('../../types')

const { SIGN_IN_KEY } = process.env

const authenticate = new mutationWithClientMutationId({
  name: 'Authenticate',
  description:
    'This mutation allows users to give their credentials and retrieve a token that gives them access to restricted content.',
  inputFields: () => ({
    authenticationCode: {
      type: GraphQLNonNull(GraphQLInt),
      description: 'Security code found in text msg, or email inbox.',
    },
    authenticateToken: {
      type: GraphQLNonNull(GraphQLString),
      description: 'The JWT that is retrieved from the sign in mutation.',
    },
  }),
  outputFields: () => ({
    authResult: {
      type: authResultType,
      description: 'The authenticated users information, and JWT.',
      resolve: async (payload) => {
        return payload.authResult
      },
    },
  }),
  mutateAndGetPayload: async (
    args,
    {
      i18n,
      query,
      auth: { tokenize, verifyToken },
      loaders: { userLoaderByKey },
      validators: { cleanseInput },
    },
  ) => {
    // Cleanse Inputs
    const authenticationCode = args.authenticationCode
    const authenticationToken = cleanseInput(args.authenticateToken)

    // Gather token parameters
    const tokenParameters = verifyToken({
      token: authenticationToken,
      secret: String(SIGN_IN_KEY),
    })

    if (
      tokenParameters.userKey === 'undefined' ||
      typeof tokenParameters.userKey === 'undefined'
    ) {
      console.warn(`Authentication token does not contain the userKey`)
      throw new Error(i18n._(t`Unable to authenticate. Please try again.`))
    }

    // Gather sign in user
    const user = await userLoaderByKey.load(tokenParameters.userKey)

    if (typeof user === 'undefined') {
      console.warn(
        `User: ${tokenParameters.userKey} attempted to authenticate, no account is associated with this id.`,
      )
      throw new Error(i18n._(t`Unable to authenticate. Please try again.`))
    }

    // Check to see if security token matches the user submitted one
    if (authenticationCode === user.tfaCode) {
      const token = tokenize({ parameters: { userKey: user._key } })

      // Reset Failed Login attempts
      try {
        await query`
                FOR u IN users
                  UPDATE ${user._key} WITH { tfaCode: null } IN users
              `
      } catch (err) {
        console.error(
          `Database error ocurred when resetting failed attempts for user: ${user._key} during authentication: ${err}`,
        )
        throw new Error(i18n._(t`Unable to authenticate. Please try again.`))
      }

      console.info(
        `User: ${user._key} successfully authenticated their account.`,
      )
      user.id = user._key

      return {
        authResult: {
          token,
          user,
        },
      }
    } else {
      console.warn(
        `User: ${user._key} attempted to authenticate their account, however the tfaCodes did not match.`,
      )
      throw new Error(i18n._(t`Unable to authenticate. Please try again.`))
    }
  },
})

module.exports = {
  authenticate,
}
